### PR TITLE
Add Discontinuous Galerkin Partial Assembly support for NCMesh.

### PR DESCRIPTION
**Pull request by YohannDudouit**
This PR adds:
- Support for Discontinuous Galerkin methods using Partial Assembly on `NCMesh`.
- A more user friendly method in `Mesh` to access face information: `Mesh::GetFaceInformation`

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#2195](https://github.com/mfem/mfem/pull/2195) | @YohannDudouit | @v-dobrev | @artv3 + @rw-anderson + @v-dobrev | 10/14/21 | 02/10/22 | 02/16/22 | |

### CODE REVIEW

File: mesh/mesh.cpp [hunk 1]
```diff
@@ -1041,6 +1041,63 @@ FaceElementTransformations *Mesh::GetBdrFaceTransformations(int BdrElemNo)
    return tr;
 }
 
+void Mesh::GetFaceInformation(int f, FaceInformation &info) const
```
Comment (pazner) [inline mesh/mesh.cpp:1044]: This can probably return by value instead of modify the info passed by reference.

**Comment by artv3** [comment]
I seem to get a lot of warnings from this branch: 

```
mesh/mesh.hpp(1330): warning: missing return statement at end of non-void function "mfem::Mesh::FaceInformation::IsOfFaceType"

mesh/../fem/restriction.hpp(523): warning: __host__ annotation is ignored on a function("InterpConfig") that is explicitly defaulted on its first declaration

mesh/../fem/restriction.hpp(523): warning: __device__ annotation is ignored on a function("InterpConfig") that is explicitly defaulted on its first declaration

mesh/../fem/restriction.hpp(526): warning: __host__ annotation is ignored on a function("operator=") that is explicitly defaulted on its first declaration

mesh/../fem/restriction.hpp(526): warning: __device__ annotation is ignored on a function("operator=") that is explicitly defaulted on its first declaration

mesh/mesh.hpp(1330): warning: missing return statement at end of non-void function "mfem::Mesh::FaceInformation::IsOfFaceType"


```

File: fem/bilininteg_dgtrace_pa.cpp [hunk 2]
```diff
@@ -182,22 +182,22 @@ void DGTraceIntegrator::SetupPA(const FiniteElementSpace &fes, FaceType type)
...
+         if (face.IsLocal() && face.IsNonConformingMaster())
+         {
+            // We skip local non-conforming master faces
+            continue;
+         }
+         else if ( face.IsOfFaceType(type) )
          {
             FaceElementTransformations &T =
-               *fes.GetMesh()->GetFaceElementTransformations(f);
+               *fes.GetMesh()->GetFaceElementTransformations(f,5);
```
[the remark this answers is not in the document]
Comment (YohannDudouit) [reply fem/bilininteg_dgtrace_pa.cpp]: done.

[the remark this answers is not in the document]
Comment (pazner) [reply fem/bilininteg_dgtrace_pa.cpp]: Might want to use `FaceElementTransformations::HAVE_ELEM1 & FaceElementTransformations::HAVE_LOC1` instead. More verbose, but also more self-documenting.

File: fem/restriction.hpp [hunk 3]
```diff
@@ -172,121 +172,590 @@ class H1FaceRestriction : public FaceRestriction
...
+                                            const int face_index);
 };
 
-// Return the face degrees of freedom returned in Lexicographic order.
+/** This class stores which side is the master non-conforming side and the
+    index of the interpolator, see InterpolationManager class below. */
+struct InterpConfig
+{
+   int config;
+   enum { conforming = -1 }; // helper value
```
[the remark this answers is not in the document]
Comment (v-dobrev) [reply fem/restriction.hpp:477]: I don't have a special preference here.

File: mesh/mesh.hpp [hunk 4]
```diff
@@ -880,13 +938,25 @@ class Mesh
...
        Similarly, if type==Interior, the "fake" boundary faces associated to
        visualization are counted as interior faces. */
-   int GetNFbyType(FaceType type) const;
+   virtual int GetNFbyType(FaceType type) const;
+
+   /** @brief Return the PointMatrix of NCFace index @a i. */
+   const DenseMatrix* GetNCFacesPtMat(int i) const
+   {
+      return nc_faces_info[i].PointMatrix;
+   }
```
[the remark this answers is not in the document]
Comment (v-dobrev) [reply mesh/mesh.hpp:959]: I think either way will work, go with your preference.

File: fem/bilininteg_dgtrace_pa.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (v-dobrev) [reply fem/bilininteg_dgtrace_pa.cpp]: You should use "`|`" with the enum constants, not "`&`".

[the remark this answers is not in the document]
Comment (pazner) [reply fem/bilininteg_dgtrace_pa.cpp]: Oops. Yes, definitely. My bad. 🤦‍♂️

**Comment by YohannDudouit** [comment]
The (hopefully) final bug has been found! Thanks a lot to @camierjs @v-dobrev and @pazner for their help on this PR!

File: fem/bilininteg_dgtrace_pa.cpp [hunk 5]
```diff
@@ -182,22 +182,22 @@ void DGTraceIntegrator::SetupPA(const FiniteElementSpace &fes, FaceType type)
...
+         if (face.IsLocal() && face.IsNonConformingMaster())
+         {
+            // We skip local non-conforming master faces
+            continue;
+         }
+         else if ( face.IsOfFaceType(type) )
          {
             FaceElementTransformations &T =
-               *fes.GetMesh()->GetFaceElementTransformations(f);
+               *fes.GetMesh()->GetFaceElementTransformations(f,5); // Elem1 and Loc1
```
Comment (rw-anderson) [inline fem/bilininteg_dgtrace_pa.cpp:196]: Can 5 use a symbol to be more self-describing? Maybe from an enum?

File: general/hash.hpp [hunk 6]
```diff
@@ -188,11 +188,11 @@ class HashTable : public BlockArray<T>
    Array<int> unused;
 
    // hash functions (NOTE: the constants are arbitrary)
-   inline int Hash(int p1, int p2) const
-   { return (984120265*p1 + 125965121*p2) & mask; }
+   inline int Hash(size_t p1, size_t p2) const
+   { return (984120265ul*p1 + 125965121ul*p2) & mask; }
```
Comment (rw-anderson) [inline general/hash.hpp:192]: Mysterious. I don't know how to improve it, though.

File: mesh/ncmesh.cpp [hunk 7]
```diff
@@ -1671,43 +1690,55 @@ void NCMesh::DerefineElement(int elem)
...
+      for (int i = 0; i < nb_square_childs; i++)
       {
-         Element &ch = elements[child[quad_deref_table[rt1][i]]];
+         const int child_local_index = quad_deref_table[ref_type_key][i];
+         const int child_global_index = child[child_local_index];
+         Element &ch = elements[child_global_index];
          el.node[i] = ch.node[i];
       }
-      for (int i = 0; i < 4; i++)
+      constexpr int nb_square_face = 4;
```
Comment (rw-anderson) [inline mesh/ncmesh.cpp:1708]: very minor, but maybe plural for consistency with other similar constants.

**Review by rw-anderson — APPROVED** [review]
I like the self-documenting/readability improvements. It looks good to me.

File: fem/bilininteg_dgtrace_pa.cpp (hunk 5 above)
Comment (pazner) [reply fem/bilininteg_dgtrace_pa.cpp:196]: https://github.com/mfem/mfem/pull/2195#discussion_r745072458

File: general/hash.hpp (hunk 6 above)
Comment (YohannDudouit) [reply general/hash.hpp:192]: It's a change that prevents overflow warnings when debugging.

File: mesh/ncmesh.cpp (hunk 7 above)
Comment (YohannDudouit) [reply mesh/ncmesh.cpp]: Thank you, I changed `face` for `faces`.

File: fem/bilininteg_dgtrace_pa.cpp (hunk 5 above)
Comment (YohannDudouit) [reply fem/bilininteg_dgtrace_pa.cpp]: I used a `const int mask` variable to make the code more self documented.

File: fem/bilininteg_dgtrace_pa.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (YohannDudouit) [reply fem/bilininteg_dgtrace_pa.cpp]: done.

File: mesh/mesh.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (YohannDudouit) [reply mesh/mesh.hpp]: done.

File: fem/restriction.cpp [hunk 8]
```diff
@@ -558,20 +559,21 @@ void L2ElementRestriction::FillJAndData(const Vector &ea_data,
    });
 }
 
-// Return the face degrees of freedom returned in Lexicographic order.
+/** Return the face degrees of freedom returned in Lexicographic order.
+    Note: Only for quad and hex */
 void GetFaceDofs(const int dim, const int face_id,
-                 const int dof1d, Array<int> &faceMap)
+                 const int dof1d, Array<int> &face_map)
```
[the remark this answers is not in the document]
Comment (YohannDudouit) [reply fem/restriction.cpp:565]: I tried implementing this, the problem is that `dof1d` is a runtime parameter.

File: fem/restriction.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (YohannDudouit) [reply fem/restriction.hpp]: done.

**Review by artv3 — APPROVED** [review]
Looks good to me! I just see one failing check:

 repo-check / branch-history (push) Failing after 16s — branch-history 

but once that gets fixed should be good to go!

**Comment by tzanio** [comment]
The `branch-history` failure is because this PR has more than 200 commits. We can ignore it for now.

File: mesh/mesh.hpp [hunk 9]
```diff
@@ -91,6 +91,64 @@ class Mesh
...
+       the master element as Elem2. On the other side, non-conforming master
+       faces always have the master element as Elem1, and one of the slave
+       element as Elem2. Except for ghost non-conforming slave faces, where
+       Elem1 is the master side and Elem2 is the slave side.
+
+       The indices of Elem1 and Elem2 can be indirectly extracted from
+       FaceInfo::Elem1No and FaceInfo::Elem2No, read the note below for special
+       cases on the index of Elem2.
+
+       The local face identifiants are deciphered from FaceInfo::Elem1Inf and
```
Comment (dylan-copeland) [inline mesh/mesh.hpp:123]: Nice documentation, but identifiant is not an English word. Do you mean index or identity?

Comment (YohannDudouit) [reply mesh/mesh.hpp:123]: ID

File: mesh/mesh.cpp [hunk 10]
```diff
@@ -1119,6 +1119,100 @@ int Mesh::GetBdrFace(int BdrElemNo) const
...
+            }
+         }
+         else // ncface >= 0
+         {
+            if (inf2 < 0)
+            {
+               face.location = e2==-1 ? FaceLocation::Local : FaceLocation::Shared;
+               face.conformity = FaceConformity::NonConformingMaster;
+               face.elem_2_index = e2==-1 ? e2 : -1 - e2;
+               face.elem_2_orientation = inf2%64;
```
Comment (YohannDudouit) [inline mesh/mesh.cpp:1182]: @v-dobrev This case seems to be problematic for @dylan-copeland, and seems to return incorrect information, the documentation states:
```
   // - Elem2No < 0 --> local "boundary" face; can be one of:
   //    - NCFace >= 0 --> non-conforming face; can be one of:
   //       - Elem2Inf < 0 --> master non-conforming face, interior or shared;
   //         In this case, Elem2No is -1; see GenerateNCFaceInfo().
```
Do you know if this code decipher the face information correctly?

Comment (dylan-copeland) [reply mesh/mesh.cpp:1182]: For a shared master NC face, it is unclear how to determine if it is actually shared or on the boundary, since `Elem2No = -1`.

**Comment by YohannDudouit** [comment]
@v-dobrev Did you want to take a final look or can we merge this?

**Review by v-dobrev — APPROVED** [review]

**Comment by v-dobrev** [comment]
Merged in `next` for testing...

**Comment by v-dobrev** [comment]
@YohannDudouit, thanks for updating the unit tests.

**Pull request merged**